From 119aa34537bf1a1f38aa2d62cea52b9937030e90 Mon Sep 17 00:00:00 2001
From: Subbaraya Sundeep <sbhatta@marvell.com>
Date: Tue, 9 Mar 2021 15:52:54 +0530
Subject: [PATCH 1306/1921] octeontx2: Re-enable FLR and MBOX interrupts

When HWVF_RST for a VF is done by AF driver in
FLR handler then the VFs parent PF driver need to
re-enable FLR and MBOX interrupts of that VF.
This is because VF related registers in PF
register space are also cleared during HWVF_RST.

Fixes: 259c3ba4("octeontx2-af: Reset MSIX config space in FLR")
Change-Id: I3943223bbce0598c38e95136d6c2a9159608ca75
Signed-off-by: Subbaraya Sundeep <sbhatta@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/47502
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/soc/marvell/octeontx2-npa/npa.c    | 9 +++++++++
 drivers/soc/marvell/octeontx2-rm/otx2_rm.c | 9 +++++++++
 drivers/soc/marvell/octeontx2-sdp/sdp.c    | 9 +++++++++
 3 files changed, 27 insertions(+)

diff --git a/drivers/soc/marvell/octeontx2-npa/npa.c b/drivers/soc/marvell/octeontx2-npa/npa.c
index 08e0c186a653..f996cf0190b2 100644
--- a/drivers/soc/marvell/octeontx2-npa/npa.c
+++ b/drivers/soc/marvell/octeontx2-npa/npa.c
@@ -466,6 +466,15 @@ static void npa_pfvf_flr_handler(struct work_struct *work)
 	otx2_enable_afpf_mbox_intr(npa);
 	npa_write64(npa, BLKADDR_RVUM, 0, RVU_PF_VFTRPENDX(vf->vf_id / 64),
 		    BIT_ULL(vf->intr_idx));
+
+	/* Re-enable MBOX and FLR interrupt as it gets cleared
+	 * in HWVF_RST reset
+	 */
+	npa_write64(npa, BLKADDR_RVUM, 0, RVU_PF_VFFLR_INT_ENA_W1SX(vf->vf_id),
+		   BIT_ULL(vf->intr_idx));
+	npa_write64(npa, BLKADDR_RVUM, 0,
+		   RVU_PF_VFPF_MBOX_INT_ENA_W1SX(vf->vf_id / 64),
+		   BIT_ULL(vf->intr_idx));
 }
 
 static void npa_pfvf_mbox_handler_up(struct work_struct *work)
diff --git a/drivers/soc/marvell/octeontx2-rm/otx2_rm.c b/drivers/soc/marvell/octeontx2-rm/otx2_rm.c
index 00845dafa9c2..6cd09421072e 100644
--- a/drivers/soc/marvell/octeontx2-rm/otx2_rm.c
+++ b/drivers/soc/marvell/octeontx2-rm/otx2_rm.c
@@ -453,6 +453,15 @@ static void rm_pfvf_flr_handler(struct work_struct *work)
 	enable_af_mbox_int(rm->pdev);
 	rm_write64(rm, BLKADDR_RVUM, 0, RVU_PF_VFTRPENDX(vf->vf_id / 64),
 		   BIT_ULL(vf->intr_idx));
+
+	/* Re-enable MBOX and FLR interrupt as it gets cleared
+	 * in HWVF_RST reset
+	 */
+	rm_write64(rm, BLKADDR_RVUM, 0, RVU_PF_VFFLR_INT_ENA_W1SX(vf->vf_id),
+		   BIT_ULL(vf->intr_idx));
+	rm_write64(rm, BLKADDR_RVUM, 0,
+		   RVU_PF_VFPF_MBOX_INT_ENA_W1SX(vf->vf_id / 64),
+		   BIT_ULL(vf->intr_idx));
 }
 
 static void rm_pfvf_mbox_handler_up(struct work_struct *work)
diff --git a/drivers/soc/marvell/octeontx2-sdp/sdp.c b/drivers/soc/marvell/octeontx2-sdp/sdp.c
index 152da151a960..699fcac2404c 100644
--- a/drivers/soc/marvell/octeontx2-sdp/sdp.c
+++ b/drivers/soc/marvell/octeontx2-sdp/sdp.c
@@ -471,6 +471,15 @@ static void sdp_pfvf_flr_handler(struct work_struct *work)
 	enable_af_mbox_int(sdp->pdev);
 	sdp_write64(sdp, BLKADDR_RVUM, 0, RVU_PF_VFTRPENDX(vf->vf_id / 64),
 		   BIT_ULL(vf->intr_idx));
+
+	/* Re-enable MBOX and FLR interrupt as it gets cleared
+	 * in HWVF_RST reset
+	 */
+	sdp_write64(sdp, BLKADDR_RVUM, 0, RVU_PF_VFFLR_INT_ENA_W1SX(vf->vf_id),
+		   BIT_ULL(vf->intr_idx));
+	sdp_write64(sdp, BLKADDR_RVUM, 0,
+		   RVU_PF_VFPF_MBOX_INT_ENA_W1SX(vf->vf_id / 64),
+		   BIT_ULL(vf->intr_idx));
 }
 
 static void sdp_pfvf_mbox_handler_up(struct work_struct *work)
-- 
2.31.1

